rs/sns/swap/src/swap.rs

```diff
@@ -1641,19 +1648,25 @@ impl Swap {
                     (None, parsed_buyer_principal, None)
                 }
                 Some(Investor::CommunityFund(CfInvestment {
+                    controller,
+                    hotkeys: _,
                     hotkey_principal,
                     nns_neuron_id,
                 })) => {
-                    let parsed_hotkey_principal = match string_to_principal(hotkey_principal) {
-                        Some(p) => p,
-                        // principal_str should always be parseable as a PrincipalId as that is enforced
-                        // in `refresh_buyer_tokens`. In the case of a bug due to programmer error, increment
-                        // the invalid field. This will require a manual intervention via an upgrade to correct
-                        None => {
-                            sweep_result.invalid += 1;
-                            continue;
-                        }
-                    };
+                    let parsed_hotkey_principal =
+                        match (controller, string_to_principal(hotkey_principal)) {
+                            (Some(p1), Some(p2)) if *p1 == p2 => *p1,
+                            // If hotkey_principal refers to a different principal than controller,
+                            // or if neither is set, something has gone wrong and we increment the invalid field.
+                            // This will require a manual intervention via an upgrade to correct.
+                            (Some(_), Some(_)) | (None, None) => {
+                                log!(ERROR, "Invalid NF neuron: recipe={recipe:?}");
+                                sweep_result.invalid += 1;
+                                continue;
+                            }
+                            (Some(p), _) => *p,
+                            (_, Some(p)) => p,
+                        };
```

Comment:
Try clarifying this code

Reply:
I ended up not being able to clarify it much with an intermediate struct, but I was able to extract the controller-getting code to a function